### PR TITLE
`File upload exercises`: Allow import of file upload exercises 

### PR DESCRIPTION
**Pull request by tobias-lippert**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
File Upload Exercises have been the only exercise type for which the import functionality was missing.
### Description

This adds two new REST endpoints in the `FileUploadExerciseResource`: ` POST file-upload-exercises/import/:id` for the actual import and `GET file-upload-exercises` to get all file-upload exercises are user has access to.
On the client side the `file-upload-exercise-update-component` is updated and the `paging-service` is introduced for file-upload exercises as well as an import method in the `file-upload-exercise-service`

### Steps for Testing

Prerequisites:
- 1 Editor or more privileged user
- 1 Course with a File Upload Exercise 
- 1 Course where the exercise is not yet part of.

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to the course where you want to import the exercise
4. Scroll down to File Upload Exercises and click Import File Upload Exercise
5. Search for the name of the exercise (or directly select it if it's shown on the first page)
6. Fill some fields on the page that opens and click Save.
7. The exercise should now be imported into the course

#### Exam Mode Testing
Doesn't affect exam mode

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

file-upload-exercise-import

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java [hunk 1]
```diff
@@ -108,6 +119,28 @@ public ResponseEntity<FileUploadExercise> createFileUploadExercise(@RequestBody
         return ResponseEntity.created(new URI("/api/file-upload-exercises/" + result.getId())).body(result);
     }
 
+    @PostMapping("/file-upload-exercises/import/{sourceId}")
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:122]:
```suggestion
    @PostMapping("file-upload-exercises/import/{sourceId}")
```

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 2]
```diff
@@ -96,6 +103,63 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
...
+                        }
+                    } else {
+                        // Lock individual mode for exam exercises
+                        this.fileUploadExercise.mode = ExerciseMode.INDIVIDUAL;
+                        this.fileUploadExercise.teamAssignmentConfig = undefined;
+                        this.fileUploadExercise.teamMode = false;
+                        // Exam exercises cannot be not included into the total score
+                        if (this.fileUploadExercise.includedInOverallScore === IncludedInOverallScore.NOT_INCLUDED) {
+                            this.fileUploadExercise.includedInOverallScore = IncludedInOverallScore.INCLUDED_COMPLETELY;
+                        }
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:139]: This should probably be moved for the server, also for the other exercise types.

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 3]
```diff
@@ -73,29 +81,64 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
...
-                    error: (error: HttpErrorResponse) => onError(this.alertService, error),
-                });
-            }
-            // Exam exercises cannot be not included into the total score
-            if (this.isExamMode && this.fileUploadExercise.includedInOverallScore === IncludedInOverallScore.NOT_INCLUDED) {
-                this.fileUploadExercise.includedInOverallScore = IncludedInOverallScore.INCLUDED_COMPLETELY;
-            }
-
-            this.saveCommand = new SaveExerciseCommand(this.modalService, this.popupService, this.fileUploadExerciseService, this.backupExercise, this.editType, this.alertService);
+            this.examCourseId = this.fileUploadExercise.course?.id || this.fileUploadExercise.exerciseGroup?.exam?.course?.id;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:84]: There should be a method like `getCourseIdViaMemberOrExerciseGroup` that does exactly that.

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 4]
```diff
@@ -131,7 +176,7 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
         this.fileUploadExercise.categories = categories;
     }
 
-    private onSaveSuccess(exercise: Exercise) {
+    onSaveSuccess(exercise: Exercise) {
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:179]: Why is this needed? I don't see this method used of this class.

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise.service.ts [hunk 5]
```diff
@@ -95,4 +96,19 @@ export class FileUploadExerciseService implements ExerciseServicable<FileUploadE
         fileUploadExercise.filePattern = fileUploadExercise.filePattern!.replace(/\s/g, '').toLowerCase();
         return fileUploadExercise;
     }
+    /**
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise.service.ts:99]:
```suggestion

    /**
```

**Review by kekskatze — DISMISSED** [review]
Tested on ts1. Everything worked as expected

**Review by LiaLism — DISMISSED** [review]
Everything works as expected. Tested on ts1.

**Review by PhilipWerz — DISMISSED** [review]
Everything works as expected. Tested on TS2.

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java [hunk 6]
```diff
@@ -108,6 +119,43 @@ public ResponseEntity<FileUploadExercise> createFileUploadExercise(@RequestBody
...
+     * @return The imported exercise (200), a not found error (404) if the template does not exist, or a forbidden error
+     *         (403) if the user is not at least an editor in the target course.
+     */
+    @PostMapping("file-upload-exercises/import/{sourceId}")
+    @PreAuthorize("hasRole('EDITOR')")
+    public ResponseEntity<FileUploadExercise> importFileUploadExercise(@PathVariable long sourceId, @RequestBody FileUploadExercise importedFileUploadExercise)
+            throws URISyntaxException {
+
+        if (sourceId <= 0 || (importedFileUploadExercise.getCourseViaExerciseGroupOrCourseMember() == null && importedFileUploadExercise.getExerciseGroup() == null)) {
+            log.debug("Either the courseId or exerciseGroupId must be set for an import");
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:142]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StaticCodeAnalysisResource.java [hunk 7]
```diff
@@ -105,13 +99,46 @@ public ResponseEntity<Set<StaticCodeAnalysisCategory>> resetStaticCodeAnalysisCa
...
+    public ResponseEntity<Set<StaticCodeAnalysisCategory>> importStaticCodeAnalysisCategoriesFromExercise(@PathVariable Long exerciseId, @RequestParam Long sourceExerciseId) {
+        log.debug("REST request to import static code analysis categories to programming exercise {} from exercise {}", exerciseId, sourceExerciseId);
+
+        ProgrammingExercise targetExercise = programmingExerciseRepository.findByIdElseThrow(exerciseId);
+        checkSCAEnabledForExerciseElseThrow(targetExercise);
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.EDITOR, targetExercise, null);
+
+        ProgrammingExercise sourceExercise = programmingExerciseRepository.findByIdElseThrow(sourceExerciseId);
+        checkSCAEnabledForExerciseElseThrow(sourceExercise);
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.EDITOR, sourceExercise, null);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StaticCodeAnalysisResource.java:128]: Optional: You loaded the user earlier for two of these checks but not here.

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 8]
```diff
@@ -73,29 +78,65 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
...
+            getCourseId(fileUploadExercise);
+            this.examCourseId = getCourseId(fileUploadExercise);
         });
 
         this.activatedRoute.queryParams.subscribe((params) => {
             if (params.shouldHaveBackButtonToWizard) {
                 this.goBackAfterSaving = true;
             }
         });
+        this.activatedRoute.url
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:90]: Can you use some helper methods to keep this section shorter and less indented?

File: src/main/webapp/app/exercises/programming/manage/grading/programming-exercise-grading-table-actions.component.ts [hunk 9]
```diff
@@ -8,28 +13,25 @@ import { ProgrammingExercise } from 'app/entities/programming-exercise.model';
...
-            class="btn btn-secondary ms-3 my-1"
-            (click)="onReset.emit()"
-            [disabled]="isSaving"
-            jhiTranslate="artemisApp.programmingExercise.configureGrading.reset"
-        ></button>
-    `,
+    templateUrl: './programming-exercise-grading-table-actions.component.html',
 })
 export class ProgrammingExerciseGradingTableActionsComponent {
+    readonly faCopy = faCopy;
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/grading/programming-exercise-grading-table-actions.component.ts:19]: We normally have the icons below all other attributes and for whatever reason not as readonly

**Review by JohannesStoehr — COMMENTED** [review]
Left some comments, code overall looks quite good

File: src/main/java/de/tum/in/www1/artemis/service/StaticCodeAnalysisService.java [hunk 10]
```diff
@@ -142,6 +142,32 @@ public Set<StaticCodeAnalysisCategory> resetCategories(ProgrammingExercise exerc
...
+     */
+    public Set<StaticCodeAnalysisCategory> importCategoriesFromExercise(ProgrammingExercise sourceExercise, ProgrammingExercise targetExercise) {
+        var sourceCategories = findByExerciseId(sourceExercise.getId());
+        var oldCategories = findByExerciseId(targetExercise.getId());
+
+        var newCategories = sourceCategories.stream().map(category -> {
+            var copy = category.copy();
+            copy.setProgrammingExercise(targetExercise);
+            return copy;
+        }).collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/StaticCodeAnalysisService.java:160]: I'd keep it like that.

From the peek-JavaDoc:
> API Note:
    This method exists mainly to support debugging, where you want to see the elements as they flow past a certain point in a pipeline

Since we change the object, we should use map() here.

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts (hunk 8 above)
Comment (tobias-lippert) [reply src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:90]: I extracted these long if else parts into seperate methods.

File: src/main/java/de/tum/in/www1/artemis/web/rest/StaticCodeAnalysisResource.java (hunk 7 above)
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StaticCodeAnalysisResource.java]: this code is not part of my PR, so I'm not sure if I should change anything here.

File: src/main/webapp/app/exercises/programming/manage/grading/programming-exercise-grading-table-actions.component.ts (hunk 9 above)
Comment (tobias-lippert) [reply src/main/webapp/app/exercises/programming/manage/grading/programming-exercise-grading-table-actions.component.ts]: this code is not part of my PR, so I'm not sure if I should change anything here.

**Review by JohannesStoehr — DISMISSED** [review]
Thanks for addressing my comments. Code looks good to me

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java [hunk 11]
```diff
@@ -110,6 +119,42 @@ public ResponseEntity<FileUploadExercise> createFileUploadExercise(@RequestBody
         return ResponseEntity.created(new URI("/api/file-upload-exercises/" + result.getId())).body(result);
     }
 
+    /**
+     * POST file-upload--exercises/import: Imports an existing file upload exercise into an existing course
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:123]:
```suggestion
     * POST file-upload-exercises/import: Imports an existing file upload exercise into an existing course
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java [hunk 12]
```diff
@@ -110,6 +119,42 @@ public ResponseEntity<FileUploadExercise> createFileUploadExercise(@RequestBody
...
+     * <p>
+     * This will import the whole exercise except for the participations and Dates.
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:125]: Is this correct? This does not use the dates of the first exercise, but I can directly set them in the client during the import.
```suggestion
     * This will import the whole exercise except for the participations and dates.
```

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 13]
```diff
@@ -73,29 +78,28 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
...
-                    error: (error: HttpErrorResponse) => onError(this.alertService, error),
-                });
-            }
-            // Exam exercises cannot be not included into the total score
-            if (this.isExamMode && this.fileUploadExercise.includedInOverallScore === IncludedInOverallScore.NOT_INCLUDED) {
-                this.fileUploadExercise.includedInOverallScore = IncludedInOverallScore.INCLUDED_COMPLETELY;
-            }
-
-            this.saveCommand = new SaveExerciseCommand(this.modalService, this.popupService, this.fileUploadExerciseService, this.backupExercise, this.editType, this.alertService);
+            getCourseId(fileUploadExercise);
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:81]: You moved the code above into a method but your are not calling it.
```suggestion
             this.isExamMode = this.fileUploadExercise.exerciseGroup !== undefined;
            handleExerciseSettings();
```

**Review by JohannesStoehr — APPROVED** [review]
Code changes after Lucas' review look good.
Sorry I missed them

**Review by timo07k — DISMISSED** [review]
Works as expected. Tested on TS1.

**Review by Strohgelaender — APPROVED** [review]

**Review by Strohgelaender — APPROVED** [review]
re-approve

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts [hunk 14]
```diff
@@ -76,29 +84,28 @@ export class FileUploadExerciseUpdateComponent implements OnInit {
...
+            this.examCourseId = getCourseId(fileUploadExercise);
         });
 
         this.activatedRoute.queryParams.subscribe((params) => {
             if (params.shouldHaveBackButtonToWizard) {
                 this.goBackAfterSaving = true;
             }
         });
+        this.activatedRoute.url
+            .pipe(tap())
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts:96]: Is that empty tap necessary?

**Review by nikirovast — DISMISSED** [review]
Tested on ts1, works!

I was able to upload a file upload exercise and participate in it.

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts (hunk 14 above)
Comment (tobias-lippert) [reply src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts]: no, it's not.

**Review by JohannesStoehr — APPROVED** [review]
Code after merge and removal of the unnecessary line looks good to me

**Review by laadvo — CHANGES_REQUESTED** [review]
Tested on ts1, there seems to be a slight problem with the title of the import page
image

I think "Edit a File Upload Exercise" should be left out, so there is only "Import a File Upload Exercise". This would make it consistent with other exercises.
example: importing a text exercise
image
Otherwise the functionality works as expected.

**Review by Strohgelaender — APPROVED** [review]

**Review by laadvo — APPROVED** [review]
The issue with the title has been fixed, approved. Tested on ts2.

**Review by nikirovast — APPROVED** [review]
Reapproval!

**Review by JohannesStoehr — APPROVED** [review]
Latest commit did not include any code problems.
Code still looks good 👍

**Review by luisa1999 — APPROVED** [review]
Code looks good to me too!
